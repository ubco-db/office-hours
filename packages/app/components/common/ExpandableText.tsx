import React, { useState } from 'react'
import PropTypes from 'prop-types'

const ExpandableText = ({ text, maxLength = 200 }) => {
  const [isExpanded, setIsExpanded] = useState(false)
  const shouldTruncate = text.length > maxLength
  const displayedText =
    isExpanded || !shouldTruncate ? text : `${text.substring(0, maxLength)}...`

  return (
    <div>
      <span className="inline-block overflow-hidden text-ellipsis">
        {displayedText}
      </span>
      {shouldTruncate && (
        <button
          className="ml-2 text-blue-500 hover:text-blue-700"
          onClick={() => setIsExpanded(!isExpanded)}
        >
          {isExpanded ? 'Less' : 'More'}
        </button>
      )}
    </div>
  )
}

ExpandableText.propTypes = {
  text: PropTypes.string.isRequired,
  maxLength: PropTypes.number,
}

export default ExpandableText
