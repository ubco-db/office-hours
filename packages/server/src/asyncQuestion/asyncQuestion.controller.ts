import {
  CreateAsyncQuestions,
  ERROR_MESSAGES,
  Role,
  AsyncQuestion,
  asyncQuestionStatus,
  UpdateAsyncQuestions,
  sendEmailAsync,
  asyncQuestionEventType,
} from '@koh/common';
import {
  Body,
  Controller,
  HttpException,
  HttpStatus,
  NotFoundException,
  Param,
  Patch,
  Post,
  Res,
  UseGuards,
} from '@nestjs/common';
import { JwtAuthGuard } from '../guards/jwt-auth.guard';
import { Roles } from '../decorators/roles.decorator';
import { User } from '../decorators/user.decorator';
import { UserModel } from '../profile/user.entity';
import { AsyncQuestionModel } from './asyncQuestion.entity';
import { asyncQuestionService } from './asyncQuestion.service';
import { CourseModel } from 'course/course.entity';
import { MailService } from 'mail/mail.service';
import { Response } from 'express';
import { AsyncQuestionVotesModel } from './asyncQuestionVotes.entity';

@Controller('asyncQuestions')
@UseGuards(JwtAuthGuard)
export class asyncQuestionController {
  constructor(
    private mailService: MailService,
    private questionService: asyncQuestionService,
  ) {}

  @Post(':qid/:vote')
  @Roles(Role.STUDENT, Role.TA, Role.PROFESSOR)
  async voteQuestion(
    @Param('qid') qid: number,
    @Param('vote') vote: number,
    @User() user: UserModel,
    @Res() res: Response,
  ): Promise<AsyncQuestion> {
    const question = await AsyncQuestionModel.findOne({
      where: { id: qid },
    });

    if (!question) {
      res
        .status(HttpStatus.NOT_FOUND)
        .send({ message: ERROR_MESSAGES.questionController.notFound });
      return;
    }

    let thisUserThisQuestionVote = await AsyncQuestionVotesModel.findOne({
      where: { userId: user.id, questionId: qid },
    });

    const hasVoted = thisUserThisQuestionVote !== undefined;
    const sumVotes = thisUserThisQuestionVote?.vote ?? 0;

    const newValue = sumVotes + vote;

    const canVote = newValue <= 1 && newValue >= -1;
    if (canVote) {
      if (hasVoted) {
        thisUserThisQuestionVote.vote = newValue;
      } else {
        thisUserThisQuestionVote = new AsyncQuestionVotesModel();
        thisUserThisQuestionVote.user = user;
        thisUserThisQuestionVote.question = question;
        thisUserThisQuestionVote.vote = newValue;
      }
    }

    await thisUserThisQuestionVote.save();

    const updatedQuestion = await AsyncQuestionModel.findOne({
      where: { id: qid },
    });

    res.status(HttpStatus.OK).send({
      questionSumVotes: updatedQuestion.votesSum,
      vote: thisUserThisQuestionVote?.vote ?? 0,
    });

    return;
  }

  @Post(':cid')
  @Roles(Role.STUDENT)
  async createQuestion(
    @Body() body: CreateAsyncQuestions,
    @Param('cid') cid: number,
    @User() user: UserModel,
    @Res() res: Response,
  ): Promise<any> {
    const c = await CourseModel.findOne({
      where: { id: cid },
    });

    if (!c) {
      res
        .status(HttpStatus.NOT_FOUND)
        .send({ message: ERROR_MESSAGES.questionController.notFound });
      return;
    }
    try {
      const question = await AsyncQuestionModel.create({
        courseId: cid,
        creator: user,
        creatorId: user.id,
        course: c,
        questionAbstract: body.questionAbstract,
        questionText: body.questionText || null,
        answerText: body.answerText || null,
        aiAnswerText: body.aiAnswerText,
        questionTypes: body.questionTypes,
        status: asyncQuestionStatus.Waiting,
        visible: body.visible || false,
        createdAt: new Date(),
      }).save();
      res.status(HttpStatus.CREATED).send(question);
      return;
    } catch (err) {
      console.error(err);
      res
        .status(HttpStatus.INTERNAL_SERVER_ERROR)
        .send({ message: ERROR_MESSAGES.questionController.saveQError });
      return;
    }
  }

  @Patch(':questionId')
  async updateQuestion(
    @Param('questionId') questionId: number,
    @Body() body: UpdateAsyncQuestions,
  ): Promise<AsyncQuestion> {
    const question = await AsyncQuestionModel.findOne({
      where: { id: questionId },
      relations: ['creator', 'images'],
    });
    if (question === undefined) {
      throw new NotFoundException();
    }

    question.aiAnswerText = body.aiAnswerText;
    question.answerText = body.answerText;

    //If not creator, check if user is TA/PROF of course of question

    Object.assign(question, body);
    question.closedAt = new Date();
    question.save().then(async () => {
      //send notification
      const receiver = await UserModel.findOne({
        where: {
          id: question.creatorId,
        },
      });
      if (!receiver) {
        throw NotFoundException;
      }
      const post: sendEmailAsync = {
        receiver: receiver.email,
        subject: 'UBC helpme Async question status change',
        type: asyncQuestionEventType.deleted,
      };
      if (body.status === asyncQuestionStatus.Resolved) {
        post.type = asyncQuestionEventType.answered;
        this.mailService.sendEmail(post);
      } else if (body.status === asyncQuestionStatus.TADeleted) {
        post.type = asyncQuestionEventType.deleted;
        this.mailService.sendEmail(post);
      }
    });
    return question;
  }
}
