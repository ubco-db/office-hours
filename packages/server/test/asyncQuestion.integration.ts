import { setupIntegrationTest } from './util/testUtils';
import { asyncQuestionModule } from '../src/asyncQuestion/asyncQuestion.module';
import {
  AsyncQuestionFactory,
  CourseFactory,
  UserFactory,
  VotesFactory,
} from './util/factories';
import { asyncQuestionStatus } from '@koh/common';

describe('Async Question Integration', () => {
  const supertest = setupIntegrationTest(asyncQuestionModule);

  describe('POST /asyncQuestions/:qid/:vote', () => {
    it('should vote on a question', async () => {
      const student = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/${question.id}/1`,
      );

      expect(response.status).toBe(200);
      expect(response.body.vote).toBe(1);
    });

    it('should update an existing vote on a question', async () => {
      const student = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      await VotesFactory.create({ userId: student.id, vote: 1, question });

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/${question.id}/-1`,
      );

      expect(response.status).toBe(200);
      expect(response.body.vote).toBe(0);
    });

    it('should not allow voting beyond the allowed range', async () => {
      const student = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      await VotesFactory.create({ userId: student.id, vote: 1, question });

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/${question.id}/2`,
      );

      expect(response.status).toBe(200);
      expect(response.body.vote).toBe(1); // original vote
    });

    it('should not allow voting by unauthorized users', async () => {
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });

      const response = await supertest().post(
        `/asyncQuestions/${question.id}/1`,
      );

      expect(response.status).toBe(401);
    });

    it('should return an error if the question does not exist', async () => {
      const student = await UserFactory.create();

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/9999/1`,
      );

      expect(response.status).toBe(404);
    });
  });

  describe('POST /asyncQuestions/:cid', () => {
    it('should create a question', async () => {
      const student = await UserFactory.create();
      const course = await CourseFactory.create();
      const response = await supertest({ userId: student.id })
        .post(`/asyncQuestions/${course.id}`)
        .send({
          questionAbstract: 'abstract',
          questionText: 'text',
          answerText: 'answer',
          aiAnswerText: 'ai',
          questionTypes: [],
          visible: true,
        });

      expect(response.status).toBe(201);
      expect(response.body.aiAnswerText).toBe('ai');
      expect(response.body.answerText).toBe('answer');
      expect(response.body.status).toBe(asyncQuestionStatus.Waiting);
    });

    it('should not create a question in a non-existent course', async () => {
      const student = await UserFactory.create();
      const response = await supertest({ userId: student.id })
        .post(`/asyncQuestions/9999`)
        .send({
          questionAbstract: 'abstract',
          questionText: 'text',
          answerText: 'answer',
          aiAnswerText: 'ai',
          questionTypes: [],
          visible: true,
        });

      expect(response.status).toBe(404);
    });

    it('should not allow unauthorized users to create a question', async () => {
      const course = await CourseFactory.create();
      const response = await supertest()
        .post(`/asyncQuestions/${course.id}`)
        .send({
          questionAbstract: 'abstract',
          questionText: 'text',
          answerText: 'answer',
          aiAnswerText: 'ai',
          questionTypes: [],
          visible: true,
        });

      expect(response.status).toBe(401);
    });
  });

  describe('PATCH /asyncQuestions/:questionId', () => {
    it('should successfully update a question', async () => {
      const user = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        creatorId: user.id,
      });
      const updatedFields = {
        aiAnswerText: 'updated AI answer',
        answerText: 'updated answer',
        status: asyncQuestionStatus.Resolved,
      };

      const response = await supertest({ userId: user.id })
        .patch(`/asyncQuestions/${question.id}`)
        .send(updatedFields);

      expect(response.status).toBe(200);
      expect(response.body.aiAnswerText).toBe(updatedFields.aiAnswerText);
      expect(response.body.answerText).toBe(updatedFields.answerText);
      expect(response.body.status).toBe(updatedFields.status);
    });

    it('should return an error if the question does not exist', async () => {
      const user = await UserFactory.create();
      const nonExistentQuestionId = 9999;
      const updatedFields = {
        aiAnswerText: 'updated AI answer',
        answerText: 'updated answer',
      };

      const response = await supertest({ userId: user.id })
        .patch(`/asyncQuestions/${nonExistentQuestionId}`)
        .send(updatedFields);

      expect(response.status).toBe(404);
    });
  });
});
