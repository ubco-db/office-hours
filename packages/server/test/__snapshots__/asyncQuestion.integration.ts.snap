// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async Question Integration POST /asyncQuestions/:cid should create a question 1`] = `
Object {
  "aiAnswerText": "ai",
  "answerText": "answer",
  "closedAt": null,
  "course": Object {
    "asyncQuestionDisplayTypes": Array [],
    "coordinator_email": null,
    "courseInviteCode": "invite-code",
    "deletedAt": null,
    "enabled": true,
    "icalURL": "https://calendar.google.com/calendar/ical/t6lu2pic7u9otrbpkuk26sl34g%40group.calendar.google.com/public/basic.ics",
    "id": 1,
    "name": "CS 304",
    "questionTimer": null,
    "sectionGroupName": "CS 304",
    "selfEnroll": false,
    "semesterId": 1,
    "timezone": null,
    "zoomLink": null,
  },
  "courseId": 1,
  "createdAt": "2024-03-10T18:44:34.077Z",
  "creator": Object {
    "accountDeactivated": false,
    "accountType": "legacy",
    "defaultMessage": null,
    "desktopNotifsEnabled": false,
    "email": "user@ubc.ca",
    "firstName": "User",
    "hideInsights": Array [],
    "id": 1,
    "includeDefaultMessage": true,
    "lastName": "Person",
    "password": null,
    "phoneNotifsEnabled": false,
    "photoURL": null,
    "sid": null,
    "userRole": "user",
  },
  "creatorId": 1,
  "id": 1,
  "questionAbstract": "abstract",
  "questionText": "text",
  "questionTypes": Array [],
  "status": "Waiting",
  "taHelpedId": null,
  "visible": true,
}
`;

exports[`Async Question Integration POST /asyncQuestions/:qid/:vote should vote on a question 1`] = `
Object {
  "question": Object {
    "aiAnswerText": null,
    "answerText": null,
    "closedAt": null,
    "courseId": null,
    "createdAt": "2020-03-01T05:00:00.000Z",
    "creatorId": null,
    "id": 1,
    "questionAbstract": "question abstract",
    "questionText": "question text",
    "questionTypes": Array [],
    "status": "Resolved",
    "taHelpedId": null,
    "visible": null,
    "votes": Array [
      Object {
        "id": 1,
        "questionId": 1,
        "userId": 1,
        "vote": 1,
      },
    ],
    "votesSum": 1,
  },
  "vote": 1,
}
`;
